## Download and Clean Dataset
Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's download and read the dataset into a <em>pandas</em> dataframe.

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

### Check how many data points we have.

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.
Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.
Lets separate our data into predictors (X) and target (Y)

In [6]:
target = "Strength"
X = concrete_data.copy()
Y = X.pop(target)

In [7]:
X.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
Y.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [16]:
# number of predictors
n_cols      = X.shape[1]
input_shape = (n_cols,)

# Part A: Build a baseline model (5 marks)

Use the Keras library to build a neural network with the following:
  * One hidden layer of 10 nodes, and a ReLU activation function
  * Use the adam optimizer and the mean squared error as the loss function.

### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

In [12]:
from tensorflow import keras as K
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [13]:
# Import the train_test_split of scikit-learn
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=0.3, random_state=42)

In [18]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

### 2. Train the model on the training data using 50 epochs.

Let's call the function now to create our model.


In [20]:
model = regression_model()

Next, we will train the model for 50 epochs.

In [21]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 139101.9219
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 36975.7148
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 10132.4121
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 7482.3120
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 6689.1665
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 5935.7041
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 5250.9619
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 4616.9219
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 4032.8413
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 3517.7944
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 3045.6260
Epoch 12/50
23/23 [==============================] - 0s 1ms/step - loss: 2633.9856
Epoch 13/

### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [22]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step


180.6668701171875

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [23]:
from sklearn.metrics import mean_squared_error

In [24]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

180.66686743242903 0.0


### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.
### 5. Report the mean and the standard deviation of the mean squared errors.

Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [26]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("This is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 55.415138244628906
10/10 [==============================] - 0s 1ms/step
MSE 2: 63.74432373046875
10/10 [==============================] - 0s 1ms/step
MSE 3: 45.593570709228516
10/10 [==============================] - 0s 1ms/step
MSE 4: 49.47313690185547
10/10 [==============================] - 0s 1ms/step
MSE 5: 54.39697265625
10/10 [==============================] - 0s 960us/step
MSE 6: 59.10958480834961
10/10 [==============================] - 0s 1ms/step
MSE 7: 57.68770217895508
10/10 [==============================] - 0s 887us/step
MSE 8: 48.96474075317383
10/10 [==============================] - 0s 960us/step
MSE 9: 58.00017547607422
10/10 [==============================] - 0s 1ms/step
MSE 10: 52.387359619140625
10/10 [==============================] - 0s 1ms/step
MSE 11: 49.7962760925293
10/10 [==============================] - 0s 892us/step
MSE 12: 46.497093200683594
10/10 [==============================] - 0s 875us/step
MSE 13: 55.892578125
10/10 [=======================